### PR TITLE
Exclude 0.0.0.0 from the list of globally routable addresses

### PR DESCRIPTION
**Pull request by achanda**

### CODE REVIEW

File: src/libstd/net/ip.rs [hunk 1]
```diff
@@ -725,22 +725,22 @@ mod tests {
...
+        check(&[169, 254, 253, 242], false, false, false, true,  false,  false,    false,  false);
+        check(&[192, 0, 2, 183],     false, false, false, false, false,  false,    false,  true);
+        check(&[192, 1, 2, 183],     false, false, false, false, true,   false,    false,  false);
+        check(&[192, 168, 254, 253], false, false, true,  false, false,  false,    false,  false);
+        check(&[198, 51, 100, 0],    false, false, false, false, false,  false,    false,  true);
+        check(&[203, 0, 113, 0],     false, false, false, false, false,  false,    false,  true);
+        check(&[203, 2, 113, 0],     false, false, false, false, true,   false,    false,  false);
+        check(&[224, 0, 0, 0],       false, false, false, false, true,   true,     false,  false);
+        check(&[239, 255, 255, 255], false, false, false, false, true,   true,     false,  false);
+        check(&[255, 255, 255, 255], false, false, false, false, false,  false,    true,   false);
```
Comment (alexcrichton) [inline src/libstd/net/ip.rs]: Why the extra space added here to all columns?

**Comment by alexcrichton** [comment]
Could you also provide the rationale for this? When I was checking properties I unfortunately couldn't find a definition to what it meant to be a "global" IP, so if there's an RFC definition we can probably start stabilizing this as ell.

File: src/libstd/net/ip.rs (hunk 1 above)
Comment (achanda) [inline src/libstd/net/ip.rs]: To align with the first row, which expanded one character. If you want, I can leave this as is.

**Comment by achanda** [comment]
Global IP is taken to mean a publicly routable IP address. I did a quick search but could not find a RFC that actually defines that. My understanding is that any IP that is not reserved for a private network [1] or some special use [2] is a global IP. It might be a good idea to change the name of the function to `is_public`.

[1] https://tools.ietf.org/html/rfc1918
[2] https://tools.ietf.org/html/rfc5735

**Comment by ollie27** [comment]
The entire `0.0.0.0/8` isn't global so we should exclude all of it.

File: src/libstd/net/ip.rs [hunk 2]
```diff
@@ -131,7 +131,7 @@ impl Ipv4Addr {
     /// - test addresses used for documentation (192.0.2.0/24, 198.51.100.0/24 and 203.0.113.0/24)
```
Comment (nagisa) [inline src/libstd/net/ip.rs]: Should update this list too.

**Comment by alexcrichton** [comment]
If we're copying python in this respect then we should also at least link to the same docs

**Comment by achanda** [comment]
@alexcrichton I added a reference to the the IANA link. Addressed all other comments.

File: src/libstd/net/ip.rs (hunk 2 above)
Comment (achanda) [inline src/libstd/net/ip.rs]: Thanks, done.

File: src/libstd/net/ip.rs [hunk 3]
```diff
@@ -120,7 +120,14 @@ impl Ipv4Addr {
         self.octets()[0] == 169 && self.octets()[1] == 254
     }
 
+    /// Returns true if this address refers to source hosts on "this"
+    /// network. This property defined by RFC 5735.
+    pub fn is_this_network(&self) -> bool {
+        self.octets()[0] == 0
+    }
```
Comment (nagisa) [inline src/libstd/net/ip.rs]: Public API addition without stability annotation?

File: src/libstd/net/ip.rs [hunk 4]
```diff
@@ -120,7 +120,14 @@ impl Ipv4Addr {
...
+
     /// Returns true if the address appears to be globally routable.
+    /// http://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
```
Comment (nagisa) [inline src/libstd/net/ip.rs]: You probably want to make it a proper link? Like 

> See [iana-ipv4-special-registry](http://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml).

with

```
See [iana-ipv4-special-registry][ipv4-sr].

[ipv4-sr]: http://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
```

or something?

File: src/libstd/net/ip.rs (hunk 3 above)
Comment (achanda) [inline src/libstd/net/ip.rs]: The whole file has a
```
 11 #![unstable(feature = "ip", reason = "extra functionality has not been \
 12                                       scrutinized to the level that it should \
 13                                       be stable",
 14             issue = "27709")]
```
Isn't that enough?

Comment (nagisa) [inline src/libstd/net/ip.rs]: Right, the stability part is fine then.

**Comment by achanda** [comment]
I might just be bikeshedding, but I'm kind of not sure about the name `is_this_network`. Open to suggestions here.

File: src/libstd/net/ip.rs [hunk 5]
```diff
@@ -722,25 +732,26 @@ mod tests {
...
-        check(&[192, 1, 2, 183],     false, false, false, false, true,  false,    false,  false);
-        check(&[192, 168, 254, 253], false, false, true,  false, false, false,    false,  false);
-        check(&[198, 51, 100, 0],    false, false, false, false, false, false,    false,  true);
-        check(&[203, 0, 113, 0],     false, false, false, false, false, false,    false,  true);
-        check(&[203, 2, 113, 0],     false, false, false, false, true,  false,    false,  false);
-        check(&[224, 0, 0, 0],       false, false, false, false, true,  true,     false,  false);
-        check(&[239, 255, 255, 255], false, false, false, false, true,  true,     false,  false);
-        check(&[255, 255, 255, 255], false, false, false, false, false, false,    true,   false);
+        //    address                unspec loopbk privt  linloc global multicast brdcast doc    this_net
+        check(&[0, 0, 0, 0],         true,  false, false, false, false,  false,    false,  false, true);
```
Comment (ranma42) [inline src/libstd/net/ip.rs]: Adding the space before the `multicast` column misaligns it and all of the following.
Wouldn't it be possible to remove a space before the `false` in the first two rows and avoid changing the whitespace in the remaining part of the table?

**Comment by achanda** [comment]
@alexcrichton please let me know if this needs more work.

File: src/libstd/net/ip.rs (hunk 3 above)
Comment (alexcrichton) [inline src/libstd/net/ip.rs]: Hm we're already a little unsure about including these sorts of funcitons at all, so perhaps this can be left out for now?

Comment (achanda) [inline src/libstd/net/ip.rs]: @alexcrichton sorry, do you mean I should remove this function?

Comment (alexcrichton) [inline src/libstd/net/ip.rs]: Yeah let's remove it for now.

**Comment by achanda** [comment]
@alexcrichton does this look good now?

**Comment by alexcrichton** [comment]
@bors: r+ a67fd99460c8ab10b4c4020b5d105535c97638ff

Looks good to me, thanks!

**Pull request merged**